staging/src/k8s.io/client-go/tools/cache/shared_informer.go

```diff
@@ -255,6 +255,36 @@ func NewSharedIndexInformer(lw ListerWatcher, exampleObject runtime.Object, defa
 	return sharedIndexInformer
 }
 
+// NewSharedIndexInformer creates a new instance for the listwatcher
+// and takes a custom key function for the internal indexer.
+// The created informer will not do resyncs if the given
+// defaultEventHandlerResyncPeriod is zero.  Otherwise: for each
+// handler that with a non-zero requested resync period, whether added
+// before or after the informer starts, the nominal resync period is
+// the requested resync period rounded up to a multiple of the
+// informer's resync checking period.  Such an informer's resync
+// checking period is established when the informer starts running,
+// and is the maximum of (a) the minimum of the resync periods
+// requested before the informer starts and the
+// defaultEventHandlerResyncPeriod given here and (b) the constant
+// `minimumResyncPeriod` defined in this file.
+func NewSharedIndexInformerWithKeyFunc(lw ListerWatcher, exampleObject runtime.Object, defaultEventHandlerResyncPeriod time.Duration, indexers Indexers, keyfn KeyFunc) SharedIndexInformer {
```

Comment:
If this design is ok, I'd also like to deduplicate the code here with the one above. Technically the `NewSharedIndexInformer` function can call this one with the current KeyFunc.